AirCasting/Utils/AveragingService.swift

```diff
@@ -122,35 +125,60 @@ final class AveragingService: NSObject, ObservableObject {
        return values.map({ $0.value }).reduce(0.0, +) / Double(values.count)
     }
     
+    private func averagedMeasurementFrom(chunk: ArraySlice<MeasurementEntity>, window: AveragingWindow, measurementCount: Int) -> [MeasurementEntity] {
+        if chunk.count == measurementCount {
+            /// https://github.com/apple/swift-algorithms/blob/main/Guides/Chunked.md
+            /// For integer types, any remainder of the division is discarded so we need to add 1 to elementsInChunk/2 to get the middle value.
+            let middleIndex = chunk.index(chunk.startIndex, offsetBy: (measurementCount/2 + 1))
+            let middleMeasurement = chunk[middleIndex]
+            middleMeasurement.value = self.calculateAvg(from: chunk)
+            middleMeasurement.averagingWindow = window.rawValue
+            return [middleMeasurement]
+        } else {
+            return Array(chunk)
+        }
+    }
+    
     private func scheduleAveraging(session: SessionEntity) {
-        let uuid = session.uuid!
-        
+        guard let uuid = session.uuid else {
+            Log.info("Couldn't schedule averaging because session \(String(describing: session.name)) doesn't have UUID")
+            return
+        }
         let timer = Timer.publish(every: TimeInterval(TimeThreshold.firstThreshold.rawValue), on: .main, in: .common)
             .autoconnect()
             .first()
             .sink { [weak self] _ in
-                self?.measurementStreamStorage.accessStorage { storage in
-                    guard let session = try? storage.getExistingSession(with: uuid) else { return }
-                    self?.startPeriodicAveraging(session: session)
-                }
+                self?.startPeriodicAveraging(session: session, window: .firstThresholdWindow)
             }
         timers[uuid] = timer
     }
     
-    private func startPeriodicAveraging(session: SessionEntity) {
+    private func startPeriodicAveraging(session: SessionEntity, window: AveragingWindow) {
         guard let context = session.managedObjectContext,
               let uuid = session.uuid else { return }
-
-        let timer = Timer.publish(every: 5.0, on: .main, in: .common)
+        
+        let timer = Timer.publish(every: TimeInterval(window.rawValue), on: .main, in: .common)
             .autoconnect()
             .sink { [weak self] _ in
+                guard let self = self else { return }
                 context.perform {
                     guard let session = try? context.existingSession(uuid: uuid) else { return }
-                    self?.perform(sessionEntity: session, averagingWindow: .firstThresholdWindow)
+                    let checkWindow = self.averagingWindowFor(session: session)
+                    let windowDidChange = checkWindow != window
+                    if windowDidChange {
+                        self.startPeriodicAveraging(session: session, window: checkWindow)
```

Comment:
If windowDidChange, we are calling startPeriodicAveraging for the second time window and then windowDidChange is false, so when we call `perform` for second time window, the windowDidChange value is already false. That's why we don't calculate averaged values for already averaged measurements and they get wiped out

Reply:
You're right but it's fixable by a one-liner. The error was that I passed `window` instead of `checkWindow` to `perform`.

Reply:
Good catch :D I can't believe I missed it 🤔